# 4.7 Deriving New Variables Task

### This script contains the following points:

## Step One
#### 01. Importing Libraries
#### 02. Importing Data
#### 03. Create New Dataframe of first 1,000,000 Rows 
#### 04. Define price_label Function & Apply to Dataframe & Get Max Value
#### 05. If-Statements with the loc() Function
#### 06. If-Statements with For-Loops

## Step Two
#### 07. Create busiest_days Column & Remove busiest_day Column
#### 08. Check Values in New Column "busiest_days" & Add Observations
#### 09. Categorizing Busiest Periods of the Day in Instacart Orders
#### 10. Create a new column containing these labels called “busiest_period_of_day.”
#### 11. Checking the Frequency
#### 12. Export Dataframe as a Pickle File

# Step One

## 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing Data

In [2]:
# Turn Path into a String & Assign Python Shortcut
path = '/Users/mgottlieb/Documents/Achievement 4'

In [3]:
# Import the  ords_prods_merge.pkl dataset

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

## 03. Create New Dataframe of first 1,000,000 Rows 

In [4]:
df = ords_prods_merge[:1000000]

## 04. Define price_label Function & Apply to Dataframe & Get Max Value

In [5]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [6]:
df['price_range'] = df.apply(price_label, axis=1)

/var/folders/79/dbxtn02s0kzdngngl4z03wch0000gp/T/ipykernel_5464/2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [7]:
df['prices'].max()

14.8

## 05. If-Statements with the loc() Function

In [8]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/var/folders/79/dbxtn02s0kzdngngl4z03wch0000gp/T/ipykernel_5464/1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [9]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [10]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [11]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

## 06. If-Statements with For-Loops

In [12]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [13]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [14]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [16]:
ords_prods_merge

order_id  user_id evaluation_status  order_number  \
0          2539329        1             prior             1   
1          2398795        1             prior             2   
2           473747        1             prior             3   
3          2254736        1             prior             4   
4           431534        1             prior             5   
...            ...      ...               ...           ...   
32404854   1320836   202557             prior            17   
32404855     31526   202557             prior            18   
32404856    758936   203436             prior             1   
32404857   2745165   203436             prior             2   
32404858   3093936   205420             prior             1   

          orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                          2                  8                     7.0   
1                          3                  7                    15.0   
2                          3                 12                    21.0   
3                          4                  7                    29.0   
4                          4                 15                    28.0   
...                      ...                ...                     ...   
32404854                   2                 15                     1.0   
32404855                   5                 11                     3.0   
32404856                   2                  7                     7.0   
32404857                   3                  5                    15.0   
32404858                   4                 14                     7.0   

          days_since_prior_order_missing  product_id  add_to_cart_order  \
0                                      1         196                  1   
1                                      0         196                  1   
2                                      0         196                  1   
3                                      0         196                  1   
4                                      0         196                  1   
...                                  ...         ...                ...   
32404854                               0       43553                  2   
32404855                               0       43553                  2   
32404856                               1       42338                  4   
32404857                               0       42338                 16   
32404858                               1       28818                  8   

          reordered _merge                   product_name  aisle_id  \
0                 0   both                           Soda        77   
1                 1   both                           Soda        77   
2                 1   both                           Soda        77   
3                 1   both                           Soda        77   
4                 1   both                           Soda        77   
...             ...    ...                            ...       ...   
32404854          1   both            Orange Energy Shots        64   
32404855          1   both            Orange Energy Shots        64   
32404856          0   both          Zucchini Chips, Pesto        50   
32404857          1   both          Zucchini Chips, Pesto        50   
32404858          0   both  Hot Oatmeal Multigrain Raisin       130   

          department_id  prices    price_range_loc  
0                     7     9.0  Mid-range product  
1                     7     9.0  Mid-range product  
2                     7     9.0  Mid-range product  
3                     7     9.0  Mid-range product  
4                     7     9.0  Mid-range product  
...                 ...     ...                ...  
32404854              7     3.7  Low-range product  
32404855              7     3.7  Low-range product  
32404856             19     6.9  Mid-range product  
32404857             19     6.9  Mid-range product  
32404858       

In [17]:
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [18]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [19]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [20]:
ords_prods_merge['busiest_day'] = result

In [21]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# Step Two

## 07. Create busiest_days Column & Remove busiest_day Column

In [24]:
# Identifying the busy and slow days
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

busiest_days = [0, 1]  This would be Saturday and Sunday\
slowest_days = [4, 3]  This would be Thursday and Wednesday

In [26]:
# Creating an empty list to temorarily hold the labels  
result_second = []


# Use a loop to assign appropriate labels based based on the value of the day
for value in ords_prods_merge["orders_day_of_week"]:
    if value == 0:  
    # 0 represents Saturday, one of the busiest days
        result_second.append("Busiest days")
    elif value == 1: 
    # 1 represents Sunday, one of the busiest days
        result_second.append("Busiest days")
    elif value == 3:  
    # 3 represents Wednesday, one of the slow days
        result_second.append("Slowest days")
    elif value == 4:  
    # 4 represents Thursday, one of the slow days
        result_second.append("Slowest days")
    else:
        result_second.append("Regularly busy")

In [28]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [27]:
# Assign the list to a new column in the DataFrame
ords_prods_merge['busiest_days'] = result_second

In [29]:
# Delet old column "busiest_day" in Dataframe
del ords_prods_merge['busiest_day']

## 08. Check Values in New Column "busiest_days" & Add Observations

In [30]:
# Check Distribution of Categories in the "busiest_days" Column & How often each Label Occurs
ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

The busiest_days output shows the frequency count of each category within the column./
The data shows that the **majority of days are categorized as "Regularly busy,"** followed by a significant number of "Busiest days," and the fewest are labeled as "Slowest days." I can thusly infer that the majority of days  days have moderate activity, but there is a substantial number that have high activity, and fewer days are relatively quiet.

In [31]:
# Checking the first five rows of the ords_prods_merge, to a quick overview of its columns and initial data entries.
ords_prods_merge.head()

order_id  user_id evaluation_status  order_number  orders_day_of_week  \
0   2539329        1             prior             1                   2   
1   2398795        1             prior             2                   3   
2    473747        1             prior             3                   3   
3   2254736        1             prior             4                   4   
4    431534        1             prior             5                   4   

   order_hour_of_day  days_since_prior_order  days_since_prior_order_missing  \
0                  8                     7.0                               1   
1                  7                    15.0                               0   
2                 12                    21.0                               0   
3                  7                    29.0                               0   
4                 15                    28.0                               0   

   product_id  add_to_cart_order  reordered _merge product_name  aisle_id  \
0         196                  1          0   both         Soda        77   
1         196                  1          1   both         Soda        77   
2         196                  1          1   both         Soda        77   
3         196                  1          1   both         Soda        77   
4         196                  1          1   both         Soda        77   

   department_id  prices    price_range_loc    busiest_days  
0              7     9.0  Mid-range product  Regularly busy  
1              7     9.0  Mid-range product    Slowest days  
2              7     9.0  Mid-range product    Slowest days  
3              7     9.0  Mid-range product    Slowest days  
4              7     9.0  Mid-range product    Slowest days

I can see that "busiest_days" has been successfully added and "busiest_day" has been removed.

## 09. Categorizing Busiest Periods of the Day in Instacart Orders

In [33]:
# Printing the column names of the DataFrame to grab title of needed column
print(ords_prods_merge.columns)

Index(['order_id', 'user_id', 'evaluation_status', 'order_number',
       'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order',
       'days_since_prior_order_missing', 'product_id', 'add_to_cart_order',
       'reordered', '_merge', 'product_name', 'aisle_id', 'department_id',
       'prices', 'price_range_loc', 'busiest_days'],
      dtype='object')


In [34]:
# In looking at the columns, I see that "order_hour_of_day" will help me achieve this task
# I start ay identifying the busiest hours of the day
ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [36]:
data = {
    'order_hour_of_day': [10, 11, 14, 15, 13, 12, 16, 9, 17, 8, 18, 19, 20, 7, 21, 22, 23, 6, 0, 1, 5, 2, 4, 3],
    'count': [2761760, 2736140, 2689136, 2662144, 2660954, 2618532, 2535202, 2454203, 2087654, 1718118, 1636502, 1258305, 976156, 891054, 795637, 634225, 402316, 290493, 218769, 115700, 87961, 69375, 53242, 51281]
}

df_ords_prods_merge = pd.DataFrame(data)

# Sorting the dataframe by the 'order_hour_of_day' column
df_ords_prods_merge_sorted = df_ords_prods_merge.sort_values(by='order_hour_of_day')

# Display the sorted data
print(df_ords_prods_merge_sorted)

    order_hour_of_day    count
18                  0   218769
19                  1   115700
21                  2    69375
23                  3    51281
22                  4    53242
20                  5    87961
17                  6   290493
13                  7   891054
9                   8  1718118
7                   9  2454203
0                  10  2761760
1                  11  2736140
5                  12  2618532
4                  13  2660954
2                  14  2689136
3                  15  2662144
6                  16  2535202
8                  17  2087654
10                 18  1636502
11                 19  1258305
12                 20   976156
14                 21   795637
15                 22   634225
16                 23   402316


In looking at the data this way we can that the highest volume of orders occurs during midday, particularly between 10 AM and 11 AM. The dataset shows the highest volume of orders actually occurring between 10 AM and 3 PM (not just 10 AM to 11 AM), with each hour during this period seeing counts over 2 million. Early morning hours from 12 AM (0) to 6 AM (6) show significantly fewer orders. These hours are the least active, with the number of orders gradually increasing as the morning progresses.

In [37]:
# Checking the stats
stats = df_ords_prods_merge['order_hour_of_day'].describe()
print(stats)

count    24.000000
mean     11.500000
std       7.071068
min       0.000000
25%       5.750000
50%      11.500000
75%      17.250000
max      23.000000
Name: order_hour_of_day, dtype: float64


### Determining the Threshhold
Fewest Orders: Hours less than the 25th percentile (5.75 rounded to 6).\
Average Orders: Hours between the 25th percentile (6) and 75th percentile (17.25 rounded to 17).\
Most Orders: Hours greater than the 75th percentile (17).

In [38]:
# Create a function that uses these thresholds to assign labels based off of the determined threshhold

# Creating an empty list to temorarily hold the labels  
result_third = []


# Use a loop to assign appropriate labels based based on the value of the day
for hour in ords_prods_merge["order_hour_of_day"]: 
    if hour < 6:
        result_third.append('Fewest orders') 
    elif 6 <= hour <= 17:
        result_third.append('Average orders')
    else:
        result_third.append('Most orders')

In [39]:
result_third

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average or

## 10. Create a new column containing these labels called “busiest_period_of_day.”

In [41]:
# Add the results as a new column to the DataFrame
ords_prods_merge['busiest_period_of_day'] = result_third

In [42]:
# Checking the first five rows of the ords_prods_merge, to check to make sure new column has been added
ords_prods_merge.head()

order_id  user_id evaluation_status  order_number  orders_day_of_week  \
0   2539329        1             prior             1                   2   
1   2398795        1             prior             2                   3   
2    473747        1             prior             3                   3   
3   2254736        1             prior             4                   4   
4    431534        1             prior             5                   4   

   order_hour_of_day  days_since_prior_order  days_since_prior_order_missing  \
0                  8                     7.0                               1   
1                  7                    15.0                               0   
2                 12                    21.0                               0   
3                  7                    29.0                               0   
4                 15                    28.0                               0   

   product_id  add_to_cart_order  reordered _merge product_name  aisle_id  \
0         196                  1          0   both         Soda        77   
1         196                  1          1   both         Soda        77   
2         196                  1          1   both         Soda        77   
3         196                  1          1   both         Soda        77   
4         196                  1          1   both         Soda        77   

   department_id  prices    price_range_loc    busiest_days  \
0              7     9.0  Mid-range product  Regularly busy   
1              7     9.0  Mid-range product    Slowest days   
2              7     9.0  Mid-range product    Slowest days   
3              7     9.0  Mid-range product    Slowest days   
4              7     9.0  Mid-range product    Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders

I can see that the new column "busiest_period_of_day" has been successfully added

## 11. Checking the Frequency

In [43]:
# Check the frequency of each category in the 'busiest_period_of_day' column
frequency_counts = ords_prods_merge['busiest_period_of_day'].value_counts()

# Print the frequency counts
print(frequency_counts)

busiest_period_of_day
Average orders    26105390
Most orders        5703141
Fewest orders       596328
Name: count, dtype: int64


## 12. Export Dataframe as a Pickle File

In [45]:
# Export dataframe to pkl file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merg_busiest_times_and_days.pkl'))